docs/cudf/source/errors.md

```diff
@@ -0,0 +1,46 @@
+
+| builtin error type | semantic | example use case |
+| ------------------ | -------- | ---------------- |
+|     TypeError      | Operation is applied to unsupported objects. For objects that is not **yet** supported, use `NotImplementedError` | [binary operation on two unsupported column types](https://github.com/rapidsai/cudf/blob/7d49f75df9681dbe1653029e7d508355884a6d86/python/cudf/cudf/core/column/numerical.py#L110) |
+|     ValueError     | The value of the object does not satisfy operation requirement. Usually raising this error requires data introspection. In cases where the object only accept from a finite set of values (such as enum or string set), also use `ValueError`. | [The dataframe to describe is empty](https://github.com/rapidsai/cudf/blob/7d49f75df9681dbe1653029e7d508355884a6d86/python/cudf/cudf/core/dataframe.py#L5366) |
+|     IndexError     | Array access out of range | [Retrieving rows from a column specified by an out of bound index](https://github.com/rapidsai/cudf/blob/7d49f75df9681dbe1653029e7d508355884a6d86/python/cudf/cudf/core/column/column.py#L849-L851) |
+|     KeyError       | Mapping access with invalid key | [Retrieving rows from column specified by an invalid key](https://github.com/rapidsai/cudf/blob/7d49f75df9681dbe1653029e7d508355884a6d86/python/cudf/cudf/core/indexing.py#L177) |
+| NotImplementedError| Operation of object is planned, but not yet supported | (none) |
+|     IOError        | (TBA by cuIO) | (TBA) |
+
+Custom error types, should not be used whenever possible.
+
+Error raised by the API is considered as part of the API interface and should be documented in the docstring. For a specific invalid input, if the error raised has changed, it is considered as a `breaking` change.
+
+## Handling Exceptions Thrown by Supporting Libraries
+
+cuDF depends on external libraries to function. While cuDF does extensive tests on inputs to make sure data passed to lower level is legal, supporting libraries may also perform their own data checks. Very often such checks are redundant. Double-checking adds unecessary latency to cuDF API calls and is thus unwanted. To allow reusing error checks from supporting libraries, the following section explains how cuDF proposes to map errors from supporting libraries to cuDF errors.
+
+Note that this does not mean cuDF will skip checking errors. As supporting libraries have different performance constraints to cuDF, cuDF should perform checks when necessary. An example is that libcudf [does not introspect data](https://github.com/rapidsai/cudf/issues/5505), but cuDF should.
+
+### Python Libraries
+
+If python libraries throws a builtin error type, cuDF will surface such error.
+
+If python libraries throws a custom error type, cuDF will attempt to reinterpret it into builtin error type.
```

Comment:
We may want to establish firmer rules around what this means and how we do it. I suspect that there will be some cases where this makes sense, while in others we would want to just forward the exception along. Do you have particular exceptions in mind for this? I think we'd want to see examples to decide the appropriate behavior here.

Reply:
Yeah, it was pyarrow's error types that led me to come up with this section: https://github.com/apache/arrow/blob/master/python/pyarrow/error.pxi

Reply:
Personally I don't think we need to codify this. There may be cases where Python libraries throw builtin types, but we know the reason why and might catch and either handle or reinterpret as a different, more specific and informative error. Similarly, there may be cases when a custom error type thrown by an underlying Python library could actually signal a broader class of problems in our context, so we just pass it through transparently. As long as we are consistent with a particular API over time, I think we should give ourselves the flexibility to handle errors from underlying libraries on an as-needed basis.